File: test/canon_codec_test.dart
import 'dart:typed_data';

import 'package:comtran/comtran.dart';
import 'package:test/test.dart';

void main() {
  group('encodeCanon', () {
    test('empty deck is a bare header', () {
      expect(encodeCanon([]), [
        0x43,
        0x54,
        0x44,
        0x45,
        0x43,
        0x4B,
        0x01,
        0x00,
        0,
        0,
        0,
        0,
      ]);
    });

    test('packing golden bytes (deck-format.md §2.2)', () {
      final columns = List<int>.filled(80, 0);
      columns[0] = 0xA00; // Column 1: rows 12 and 0 (plus zero).
      columns[1] = 0x001; // Column 2: row 9.
      columns[79] = 0xFFF; // Column 80: all twelve rows.
      final Uint8List bytes = encodeCanon([CardImage.fromColumns(columns)]);
      expect(bytes.length, 12 + 120);
      expect(bytes.sublist(8, 12), [0, 0, 0, 1]);
      expect(bytes.sublist(12, 15), [0xA0, 0x00, 0x01]);
      expect(bytes.sublist(15, 129), everyElement(0));
      expect(bytes.sublist(129, 132), [0x00, 0x0F, 0xFF]);
    });
  });

  group('decodeCanon', () {
    test('round-trips a deck', () {
      final deck = [
        CardImage.blank(),
        CardImage.fromColumns(
          List<int>.generate(80, (int i) => (i * 37) & 0xFFF),
        ),
      ];
      expect(decodeCanon(encodeCanon(deck)), deck);
    });

    test('rejects a short file', () {
      expect(() => decodeCanon(Uint8List(5)), throwsA(isA<FormatException>()));
    });

    test('rejects a bad magic', () {
      final Uint8List bytes = encodeCanon([]);
      bytes[0] = 0x58;
      expect(() => decodeCanon(bytes), throwsA(isA<FormatException>()));
    });

    test('rejects an unknown version', () {
      final Uint8List bytes = encodeCanon([]);
      bytes[6] = 2;
      expect(() => decodeCanon(bytes), throwsA(isA<FormatException>()));
    });

    test('rejects nonzero flags', () {
      final Uint8List bytes = encodeCanon([]);
      bytes[7] = 1;
      expect(() => decodeCanon(bytes), throwsA(isA<FormatException>()));
    });

    test('rejects a length that does not match the card count', () {
      final Uint8List bytes = encodeCanon([CardImage.blank()]);
      expect(
        () => decodeCanon(Uint8List.sublistView(bytes, 0, bytes.length - 1)),
        throwsA(isA<FormatException>()),
      );
      final Uint8List padded = Uint8List(bytes.length + 1)..setAll(0, bytes);
      expect(() => decodeCanon(padded), throwsA(isA<FormatException>()));
    });
  });
}

File: editors/vscode-punchcard/test/canonCodec.test.js
'use strict';

const assert = require('node:assert/strict');
const test = require('node:test');

const {
  CANON_FORMAT_VERSION,
  COLUMN_COUNT,
  HEADER_LENGTH,
  RECORD_LENGTH,
  blankCard,
  decodeCanon,
  encodeCanon,
} = require('../out/canonCodec.js');

const MAGIC = Buffer.from('CTDECK', 'ascii');

test('the header carries the magic, version, flags and count', () => {
  const bytes = encodeCanon([blankCard(), blankCard(), blankCard()]);
  assert.deepEqual(Buffer.from(bytes.subarray(0, 6)), MAGIC);
  assert.equal(bytes[6], CANON_FORMAT_VERSION);
  assert.equal(bytes[7], 0);
  assert.equal(new DataView(bytes.buffer).getUint32(8, false), 3);
  assert.equal(bytes.length, HEADER_LENGTH + 3 * RECORD_LENGTH);
});

test('an empty deck is a bare header', () => {
  const bytes = encodeCanon([]);
  assert.equal(bytes.length, HEADER_LENGTH);
  assert.deepEqual(decodeCanon(bytes), []);
});

test('a blank card is 120 zero bytes', () => {
  const bytes = encodeCanon([blankCard()]);
  assert.equal(bytes.length, HEADER_LENGTH + RECORD_LENGTH);
  for (let i = HEADER_LENGTH; i < bytes.length; i++) {
    assert.equal(bytes[i], 0, `byte ${i} is not zero`);
  }
});

test('two columns pack into three bytes, most significant bit first', () => {
  const card = blankCard();
  card[0] = 0xabc;
  card[1] = 0x123;
  card[78] = 0xfff;
  card[79] = 0x001;
  const bytes = encodeCanon([card]);
  assert.equal(bytes[HEADER_LENGTH + 0], 0xab);
  assert.equal(bytes[HEADER_LENGTH + 1], 0xc1);
  assert.equal(bytes[HEADER_LENGTH + 2], 0x23);
  const last = HEADER_LENGTH + RECORD_LENGTH - 3;
  assert.equal(bytes[last + 0], 0xff);
  assert.equal(bytes[last + 1], 0xf0);
  assert.equal(bytes[last + 2], 0x01);
});

test('column order is column 1 first, column 80 last', () => {
  const card = blankCard();
  for (let i = 0; i < COLUMN_COUNT; i++) {
    card[i] = (i * 7 + 1) & 0xfff;
  }
  const [back] = decodeCanon(encodeCanon([card]));
  assert.deepEqual(Array.from(back), Array.from(card));
});

test('decode rejects a short file', () => {
  assert.throws(() => decodeCanon(new Uint8Array(11)), /shorter than/);
});

test('decode rejects a bad magic', () => {
  const bytes = encodeCanon([blankCard()]);
  bytes[0] = 0x58;
  assert.throws(() => decodeCanon(bytes), /bad magic/);
});

test('decode rejects an unknown version', () => {
  const bytes = encodeCanon([blankCard()]);
  bytes[6] = 2;
  assert.throws(() => decodeCanon(bytes), /unknown canon format version 2/);
});

test('decode rejects nonzero flags', () => {
  const bytes = encodeCanon([blankCard()]);
  bytes[7] = 1;
  assert.throws(() => decodeCanon(bytes), /reserved flags byte is nonzero/);
});

test('decode rejects a length that the count does not match', () => {
  const bytes = encodeCanon([blankCard(), blankCard()]);
  assert.throws(() => decodeCanon(bytes.subarray(0, bytes.length - 1)), /does not match/);
});

test('encode rejects a card that is not 80 columns', () => {
  assert.throws(() => encodeCanon([new Uint16Array(79)]), /exactly 80 columns/);
});

test('encode rejects a column value wider than 12 bits', () => {
  const card = blankCard();
  card[5] = 0x1000;
  assert.throws(() => encodeCanon([card]), /12 bits/);
});

test('a deck of every 12-bit pattern round-trips', () => {
  const deck = [];
  let value = 0;
  for (let c = 0; c < 52; c++) {
    const card = blankCard();
    for (let i = 0; i < COLUMN_COUNT; i++) {
      card[i] = value & 0xfff;
      value++;
    }
    deck.push(card);
  }
  const back = decodeCanon(encodeCanon(deck));
  assert.equal(back.length, deck.length);
  for (let i = 0; i < deck.length; i++) {
    assert.deepEqual(Array.from(back[i]), Array.from(deck[i]));
  }
});
